File: fragments/C/heartbeats.c
#define HEARTBEAT_LIVENESS  3       //  3-5 is reasonable
#define HEARTBEAT_INTERVAL  1000    //  msecs
#define INTERVAL_INIT       1000    //  Initial reconnect
#define INTERVAL_MAX       32000    //  After exponential backoff

...
//  If liveness hits zero, queue is considered disconnected
size_t liveness = HEARTBEAT_LIVENESS;
size_t interval = INTERVAL_INIT;

//  Send out heartbeats at regular intervals
uint64_t heartbeat_at = zclock_time () + HEARTBEAT_INTERVAL;

while (true) {
    zmq_pollitem_t items [] = { { worker,  0, ZMQ_POLLIN, 0 } };
    int rc = zmq_poll (items, 1, HEARTBEAT_INTERVAL * ZMQ_POLL_MSEC);

    if (items [0].revents & ZMQ_POLLIN) {
        //  Receive any message from queue
        liveness = HEARTBEAT_LIVENESS;
        interval = INTERVAL_INIT;
    }
    else
    if (--liveness == 0) {
        zclock_sleep (interval);
        if (interval < INTERVAL_MAX)
            interval *= 2;
        zsocket_destroy (ctx, worker);
        ...
        liveness = HEARTBEAT_LIVENESS;
    }
    //  Send heartbeat to queue if it's time
    if (zclock_time () > heartbeat_at) {
        heartbeat_at = zclock_time () + HEARTBEAT_INTERVAL;
        //  Send heartbeat message to queue
    }
}

File: fragments/Java/heartbeats.java
private static final int HEARTBEAT_LIVENESS = 3;       //  3-5 is reasonable
private static final int HEARTBEAT_INTERVAL = 1000;    //  msecs
private static final int INTERVAL_INIT = 1000;    //  Initial reconnect
private static final int INTERVAL_MAX = 32000;    //  After exponential backoff

...
//  If liveness hits zero, queue is considered disconnected
int liveness = HEARTBEAT_LIVENESS;
int interval = INTERVAL_INIT;

//  Send out heartbeats at regular intervals
long heartbeat_at = System.currentTimeMillis() + HEARTBEAT_INTERVAL;

while (true) {
    PollItem items[] = { new PollItem(worker, ZMQ.Poller.POLLIN) };
    int rc = ZMQ.poll(items, 1, HEARTBEAT_INTERVAL);

    if (items[0].isReadable()) {
        //  Receive any message from queue
        liveness = HEARTBEAT_LIVENESS;
        interval = INTERVAL_INIT;
    }
    else
    if (--liveness == 0) {
        Thread.sleep(interval);
        if (interval < INTERVAL_MAX)
            interval *= 2;
        ctx.destroySocket(worker);
        ...
        liveness = HEARTBEAT_LIVENESS;
    }
    //  Send heartbeat to queue if it's time
    if (System.currentTimeMillis() > heartbeat_at) {
        heartbeat_at = System.currentTimeMillis() + HEARTBEAT_INTERVAL;
        //  Send heartbeat message to queue
    }
}
